src/tools/aml-language-server/handlers.ts

```diff
@@ -0,0 +1,141 @@
+import * as MessageTypes from './messageTypes.js';
+import {alphaNumerics, AmlServiceContext, jsonrpc, normalizeUri, uri2path} from './util.js';
+
+// tslint:disable-next-line: no-any
+export type Handler = ((params: object, context: AmlServiceContext) => any);
+
+export const handlers: Dictionary<Handler> = {
+  initialize: (params: object, context: AmlServiceContext):
+                  MessageTypes.InitializeResult => {
+    context.rootPath =
+    // tslint:disable-next-line: no-any
+        (params as any).rootPath || uri2path((params as any).rootUri);
+    const result: MessageTypes.InitializeResult = {
+      capabilities: {
+        // Tell the client that the server works in FULL text document sync mode
+        textDocumentSync: TextDocumentSyncKind.Full,
+        hoverProvider: true,
+        signatureHelpProvider: {
+          triggerCharacters: ['(', ','],
+        },
+        definitionProvider: true,
+        typeDefinitionProvider: true,
+        referencesProvider: true,
+        documentSymbolProvider: true,
+        workspaceSymbolProvider: true,
+        xworkspaceReferencesProvider: true,
+        xdefinitionProvider: true,
+        xdependenciesProvider: true,
+        completionProvider: {
+          resolveProvider: true,
+          triggerCharacters: ['.'],
+        },
+        codeActionProvider: true,
+        renameProvider: true,
+        executeCommandProvider: {
+          commands: [],
+        },
+        xpackagesProvider: true,
+      }
+    };
+    return result;
+  },
+
+  textDocumentCompletion: async (
+      params, context: AmlServiceContext):
+      Promise<MessageTypes.CompletionItem> => {
+        const uri = (params as {textDocument: {uri: string}}).textDocument.uri;
+        context.logger.info(`Completing for : ${uri}...`);
+        return undefined;
+      },
+
+  textDocumentDidSave: async (params, context: AmlServiceContext):
+      Promise<NotificationMessage> => {
+        const uri = (params as {textDocument: {uri: string}}).textDocument.uri;
+        context.logger.info(`Handling save for: ${uri}...`);
+        return publishDiagnostics(uri, context);
+      },
+
+  textDocumentDidChange: async (params, context: AmlServiceContext):
+      Promise<NotificationMessage> => {
+        const uri = (params as {textDocument: {uri: string}}).textDocument.uri;
+        context.logger.info(`Handling changes for: ${uri}...`);
+        return publishDiagnostics(uri, context);
+      },
+
+  textDocumentDidOpen: async (params, context: AmlServiceContext):
+      Promise<NotificationMessage> => {
+        const uri = (params as {textDocument: {uri: string}}).textDocument.uri;
+        context.logger.info(`Opened: ${uri}...`);
+        return publishDiagnostics(uri, context);
+      }
+};
+
+async function publishDiagnostics(uri: string, context: AmlServiceContext):
+    Promise<NotificationMessage> {
+      const diagnosticParams: PublishDiagnosticsParams =
+          await gatherDiagnostics(uri, context);
```

Comment:
Nit: no need to line break here, we're not limited to 80 chars.

Reply:
Done